plugins/source/awspricing/resources/services.go

```diff
@@ -0,0 +1,320 @@
+	}
+
+	defer resp.Body.Close()
+	offerFiles := make(map[string]any, 0)
+
+	if err := json.NewDecoder(resp.Body).Decode(&offerFiles); err != nil {
+		return nil, err
+	}
+	links := make([]string, 0)
+	regions := offerFiles["regions"].(map[string]any)
+	for _, region := range regions {
+		region := region.(map[string]any)
+		if !validateRegionsToSync(c, region["regionCode"].(string)) {
+			continue
+		}
+		links = append(links, ROOT_URL+region["currentVersionUrl"].(string))
+	}
+	return links, nil
+}
+
+func getPricingFile(ctx context.Context, meta schema.ClientMeta, resource *schema.Resource) error {
+	link := resource.Item.(string)
+	resp, err := http.Get(link)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	rawPricingFile := make(map[string]any, 0)
+
+	if err := json.NewDecoder(resp.Body).Decode(&rawPricingFile); err != nil {
+		return err
+	}
+	pricingFile, _ := transformRawPricingFile(rawPricingFile)
+	resource.Item = pricingFile
+	return nil
+}
+
+func transformRawPricingFile(rawStruct map[string]any) (PricingFile, error) {
+	var pricingFile PricingFile
+	pricingFile.FormatVersion = rawStruct["formatVersion"].(string)
+	pricingFile.Disclaimer = rawStruct["disclaimer"].(string)
+	pricingFile.OfferCode = rawStruct["offerCode"].(string)
+	pricingFile.Version = rawStruct["version"].(string)
+
+	pubDate, err := time.Parse(time.RFC3339, rawStruct["publicationDate"].(string))
+	if err != nil {
+		return PricingFile{}, err
+	}
+	pricingFile.PublicationDate = pubDate
+
+	pricingFile.Products = extractProducts(rawStruct["products"].(map[string]any))
+
+	rawTerms := rawStruct["terms"].(map[string]any)
+	terms, err := extractAllTerms(rawTerms)
+	if err != nil {
+		return PricingFile{}, err
+	}
+	pricingFile.Terms = terms
+
+	return pricingFile, nil
+}
+
+func extractProducts(rawProducts map[string]any) []Product {
+	products := make([]Product, len(rawProducts))
+	counter := 0
+	for _, val := range rawProducts {
+		val := val.(map[string]any)
+		product := Product{
+			Sku:        val["sku"].(string),
+			Attributes: extractAttributes(val["attributes"].(map[string]any)),
+		}
+		if productFamily, ok := val["productFamily"]; ok {
+			product.ProductFamily = productFamily.(string)
+		}
+		products[counter] = product
+		counter++
+	}
+	return products
+}
+
+func extractAttributes(attributes map[string]any) map[string]string {
+	returnVal := make(map[string]string)
+	for key, val := range attributes {
+		returnVal[key] = val.(string)
+	}
+	return returnVal
+}
+
+var termTypes = []string{"OnDemand", "Reserved"}
+
+func countTermItems(termsRaw map[string]any) int {
+	counter := 0
+	for _, termType := range termTypes {
+		if _, ok := termsRaw[termType]; !ok {
+			continue
+		}
+		for _, typeTermsRaw := range termsRaw[termType].(map[string]any) {
+			counter += len(typeTermsRaw.(map[string]any))
+		}
+	}
+	return counter
+}
+
+func extractAllTerms(termsRaw map[string]any) ([]Term, error) {
+	terms := make([]Term, countTermItems(termsRaw))
```

Comment:
It's fine, but I'm not sure we're winning much, efficiency-wise, by counting the terms in order to do a single allocation, rather than starting with a small slice and appending to it as we need to. I think the latter would be simpler / less code

Reply:
I guess the only thing is that the allocated slice in this instance will be exactly the right size, whereas using `append` will cause it to be up to double the required size. So I suppose that can make quite a difference in terms of memory usage if we're talking huge slices. I'll retract my above comment 😄

Reply:
I actually started with no pre-allocation and only optimized it after realizing that `Terms` can be over 100mb of data...